torch/distributed/fsdp/_wrap_utils.py

```diff
@@ -119,3 +127,138 @@ def _warn_on_overridden_mixed_precision(
         "These modules will be wrapped as separate FSDP instacnes with mixed "
         "precision disabled."
     )
+
+
+def _validate_frozen_params(
+    root_module: nn.Module,
+    modules_to_wrap: Set[nn.Module],
+    ignored_params: Set[nn.Parameter],
+    use_orig_params: bool,
+):
+    """
+    This checks that, given ``modules_to_wrap``, each module would manage
+    parameters that are uniformly frozen or non-frozen. This uniformity
+    requirement is strict for ``use_orig_params=False`` (hard error) and highly
+    recommended for ``use_orig_params=True`` (user warning).
+    """
+    topo_sorted_named_modules = _get_topo_sorted_named_modules(root_module)
+    reverse_topo_sorted_named_modules = reversed(topo_sorted_named_modules)
+    visited_modules: Set[nn.Module] = set()
+    for module_name, module in reverse_topo_sorted_named_modules:
+        if module in modules_to_wrap:
+            param_to_fqn = _get_managed_param_to_fqn(
+                module, ignored_params, visited_modules, module_name
+            )
+            frozen_param_fqns: List[str] = []
+            frozen_param_numel = 0
+            nonfrozen_param_fqns: List[str] = []
+            nonfrozen_param_numel = 0
+            for param, fqn in param_to_fqn.items():
+                if param.requires_grad:
+                    nonfrozen_param_fqns.append(fqn)
+                    nonfrozen_param_numel += param.numel()
```

Comment:
instead of reporting numel, reporting nbytes seems better since that will adjust for differing dtype size (...well, unless there's some shenanigans where dtypes have to get bucketed together?)